File: api/services/messages.py
from typing import List, Optional
from services.llm import MarketplaceSearch

def format_telegram_message(
  user_query: str,
  result: Optional[MarketplaceSearch],
  relevant_posts: List[dict]
) -> str:
  if not result or not result.listings:
    return f"🔍 Terragora Results: {user_query}\n\nNo relevant listings found."

  paired = []
  for i, listing in enumerate(result.listings):
    post = relevant_posts[i] if i < len(relevant_posts) else {}
    paired.append((listing, post))

  paired.sort(
    key=lambda x: (x[0].deal_score is not None, x[0].deal_score or 0),
    reverse=True
  )

  lines = []

  # Header
  total = len(paired)
  lines.append(f"🔍 Terragora Results: {user_query}")
  lines.append(f"📊 Found: {total} listings | Sorted by: Best Value")
  lines.append("")
  lines.append("━━━━━━━━━━━━━━━")
  lines.append("")

  # Listings
  for i, (listing, post) in enumerate(paired):
    title = post.get("title", "Unknown Item")
    price = post.get("price", "N/A")
    url = post.get("metadata", {}).get("url", "#")

    # Normalize price
    if isinstance(price, (int, float)):
      price_str = f"Rp {int(price):,}".replace(",", ".")
    else:
      price_str = f"Rp {price}"

    lines.append(f"{i+1}. {title}")
    lines.append("")
    lines.append(f"💰 Price: {price_str}")

    if listing.location:
      lines.append(f"📍 Location: {listing.location}")

    lines.append(f"📦 Condition: {listing.condition}")

    if listing.deal_score is not None:
      lines.append(f"📈 Deal Score: {listing.deal_score}/10")

    lines.append("")
    lines.append("📝 Seller Notes:")
    for note in listing.seller_notes:
      lines.append(f"• {note}")

    lines.append("")
    lines.append(f"✅ Verdict: {listing.verdict}")
    lines.append(f"⚠️ Watch Out: {listing.watch_out or '-'}")
    lines.append("")
    lines.append(f"🔗 View Post: {url}")
    lines.append("")
    lines.append("━━━━━━━━━━━━━━━")
    lines.append("")

  return "\n".join(lines).strip()

File: internal/bot-server/messages.go
package botserver

import (
  "fmt"
  "strings"

  llm "leaddits/internal/pkg/llm"
  pg "leaddits/internal/pkg/pg"
)

type TelegramMessageBuilder struct {
  userQuery string
  paired    []PairedResult
  builder   strings.Builder
}

func NewTelegramMessageBuilder(userQuery string, result *llm.MarketplaceSearch, relevantPosts []pg.RedditPost) *TelegramMessageBuilder {
  var paired []PairedResult
  if result != nil {
    for i, listing := range result.Listings {
      if i < len(relevantPosts) {
        paired = append(paired, PairedResult{
          Listing: listing,
          Post:    relevantPosts[i],
        })
      }
    }
  }

  return &TelegramMessageBuilder{
    userQuery: userQuery,
    paired:    paired,
  }
}

func (b *TelegramMessageBuilder) Build() string {
  if len(b.paired) == 0 {
    return fmt.Sprintf("🔍 Results: %s\n\nNo relevant listings found.", b.userQuery)
  }

  b.builder.WriteString(fmt.Sprintf("🔍 Results for: %s\n", b.userQuery))
  b.builder.WriteString(fmt.Sprintf("📊 Found: %d listings\n", len(b.paired)))
  b.builder.WriteString("━━━━━━━━━━━━━━━\n\n")

  for i, item := range b.paired {
    b.addListing(i+1, item)
  }

  return strings.TrimSpace(b.builder.String())
}

func (b *TelegramMessageBuilder) addListing(index int, item PairedResult) {

  // Title
  b.builder.WriteString(fmt.Sprintf("%d. %s\n", index, item.Post.Title))

  // Adjustable: Price
  formattedPrice := FormatPrice(item.Post.Price)
  if formattedPrice != "" && !strings.EqualFold(formattedPrice, "n/a") {
    b.builder.WriteString(fmt.Sprintf("💰 Price: %s\n", formattedPrice))
  }

  // Adjustable: Seller Notes
  if len(item.Listing.SellerNotes) > 0 {
    hasContent := false
    for _, note := range item.Listing.SellerNotes {
      trimmed := strings.TrimSpace(note)
      if trimmed != "" {
        if !hasContent {
          b.builder.WriteString("\n📝 Seller Notes:\n")
          hasContent = true
        }
        b.builder.WriteString(fmt.Sprintf("• %s\n", trimmed))
      }
    }
  }

  // URL
  url := ""
  if item.Post.URL != nil {
    url = *item.Post.URL
  }
  b.builder.WriteString(fmt.Sprintf("\n🔗 View Post: %s\n", url))

  b.builder.WriteString("━━━━━━━━━━━━━━━\n\n")
}

func FormatTelegramMessage(userQuery string, result *llm.MarketplaceSearch, relevantPosts []pg.RedditPost) string {
  return NewTelegramMessageBuilder(userQuery, result, relevantPosts).Build()
}
